plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -395,14 +519,163 @@ fu_rts54hub_rtd21xx_device_write_firmware (FuDevice *device,
+	if (read_buf[0] != ISP_STATUS_IDLE_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+			     "failed project ID with error 0x%02x: ",
+			     read_buf[0]);
+		return FALSE;
+	}
+
+	/* verify project ID */
+	project_addr = fu_common_read_uint32 (read_buf + 1, G_BIG_ENDIAN);
+	project_id_count = read_buf[5];
+	write_buf[0] = ISP_BG_CMD_SYNC_IDENTIFY_CODE;
+	if (!fu_memcpy_safe (write_buf, sizeof (write_buf), 0x1, /* dst */
+			     fwbuf, fwbufsz, project_addr,	/* src */
+			     project_id_count, error)) {
+		g_prefix_error (error,
+				"failed to write project ID from 0x%04x: ",
+				project_addr);
+		return FALSE;
+	}
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   write_buf,
+						   project_id_count + 1,
+						   error)) {
+		g_prefix_error (error, "failed to send fw update start cmd: ");
+		return FALSE;
+	}
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+		return FALSE;
+
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE);
+	/* background FW update start command */
+	write_buf[0] = ISP_BG_CMD_FW_UPDATE_START;
+	fu_common_write_uint16 (write_buf + 1, ISP_BACKGROUND_DATA_BLOCKSIZE, G_BIG_ENDIAN);
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   write_buf, 3,
+						   error)) {
+		g_prefix_error (error, "failed to send fw update start cmd: ");
+		return FALSE;
+	}
+
+	/* send data */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	chunks = fu_chunk_array_new_from_bytes (fw,
+						0x00,	/* start addr */
+						0x00,	/* page_sz */
+						ISP_BACKGROUND_DATA_BLOCKSIZE);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+			return FALSE;
+
+		if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+							   UC_ISP_SLAVE_ADDR,
+							   UC_BACKGROUND_ISP_DATA_OPCODE,
+							   fu_chunk_get_data (chk),
+							   fu_chunk_get_data_sz (chk),
+							   error)) {
+			g_prefix_error (error,
+					"failed to write @0x%04x: ",
+					fu_chunk_get_address (chk));
+			return FALSE;
+		}
+
+		/* update progress */
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len);
+	}
+
+	/* update finish command */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	if (!fu_rts54hub_rtd21xx_device_read_status (self, NULL, error))
+		return FALSE;
+
+	write_buf[0] = ISP_BG_CMD_FW_UPDATE_ISP_DONE;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_ISP_SLAVE_ADDR,
+						   UC_BACKGROUND_OPCODE,
+						   write_buf, 1,
+						   error)) {
+		g_prefix_error (error, "failed update finish cmd: ");
+		return FALSE;
+	}
+
+	/* exit background-fw mode */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_progress (device, 0);
```

Comment:
shouldn't this go to 100% at the end?

Reply:
Yes ，it should be 100%.

Reply:
hi @superm1 ， 
for Background fwupdate, it won’t reset after finish wirting fw. The new FW will not run until it is replug again. 
Shoud I Remove the “fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);” and add 	“fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);”

Reply:
Wait for replug will mean that it's going to disappear and reappear on it's own.  Does the device reboot on it's own, or it's next power cycle?

Reply:
Next power cycle. The device will not reboot on it's own.